elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/ResponseHandler.scala

```diff
@@ -0,0 +1,67 @@
+package com.sksamuel.elastic4s.http
+
+import java.nio.charset.Charset
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+
+object JacksonSupport {
+
+  val mapper: ObjectMapper with ScalaObjectMapper = new ObjectMapper with ScalaObjectMapper
+  mapper.registerModule(DefaultScalaModule)
+
+  mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)
+  mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
+  mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
+  mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
+}
+
+import org.apache.http.HttpEntity
+import org.elasticsearch.client.{Response, ResponseException}
+
+import scala.io.{Codec, Source}
+import scala.util.{Failure, Try}
+
+trait ResponseHandler[U] {
+  def onResponse(response: Response): Try[U]
+  def onError(e: Exception): Try[U]
+}
+
+// a ResponseHandler that marshalls the body into the required type using Jackson
+// the response body is converted into a string using a codec derived from the content encoding header
+// if the content encoding header is null, then UTF-8 is assumed
+object ResponseHandler {
```

Comment:
I guess this scaladoc belongs to the `ResponseHandler` trait and should  be written as scaladoc.

Even this is used internally, it should be more documented on _why_ it exists rather then implementation details like "utf-8" is used if missing. If I understand correctly from the other changes a `ResponseHandler` defines the behaviour regarding the http status codes. 

This trait is very similiar to [twitter finagle's response classifier](https://twitter.github.io/finagle/guide/Clients.html#response-classification). However the response classification is _only_ for classification, not for parsing, which I think is a good decision. The type definition is

```scala
type ResponseClassifier = ParticalFunction[ReqRep, ResponseClass]
```

In our case this could be something like

```scala
type ResponseClassifier = ParticalFunction[ResponseException, Try[U]]
```

However using the exception directly isn't a very good api design. Ideally the http response itself can be used. Not sure if `204` No Content can be an error as well. And what about `5xx` codes? When you use `failure404` does this exclude `5xx` codes? I'm asking, because this maybe confusing for other contributors at first sight.

Not sure if this is a good fit for the apache http implementation if it works with exception for propagating non `2xx` status codes. Still it works better for scala devs I guess, if you have meaningful types to work with,